### ISSUE TITLE
-Wdangling-false warning 

### ISSUE THREAD

**Issue by adambadura** [google/googletest#2390]
`-Wdangling-false` warning shows when building with GCC. For example:
```plain
Scanning dependencies of target googletest-port-test
[ 38%] Building CXX object googletest/CMakeFiles/googletest-port-test.dir/test/googletest-port-test.cc.o
/var/fpwork/abadura/adambadura-googletest/googletest/test/googletest-port-test.cc: In member function ‘virtual void testing::internal::GtestCheckSyntaxTest_BehavesLikeASingleStatement_Test::TestBody()’:
/var/fpwork/abadura/adambadura-googletest/googletest/test/googletest-port-test.cc:223:6: warning: suggest explicit braces to avoid ambiguous ‘else’ [-Wdangling-else]
   if (AlwaysFalse())
      ^

```

It seems all cases are associated with the use of `GTEST_CHECK_` macro defined as follows (`gtest-port.h`):
```c++
# define GTEST_CHECK_(condition) \
    GTEST_AMBIGUOUS_ELSE_BLOCKER_ \
    if (::testing::internal::IsTrue(condition)) \
      ; \
    else \
      GTEST_LOG_(FATAL) << "Condition " #condition " failed. "
#endif  // !defined(GTEST_CHECK_)
```

It does use `GTEST_AMBIGUOUS_ELSE_BLOCKER_ ` to protect precisely against this warning. However, `GTEST_AMBIGUOUS_ELSE_BLOCKER_ ` is defined as follows (also `gtest-port.h`):
```c++
#ifdef __INTEL_COMPILER
# define GTEST_AMBIGUOUS_ELSE_BLOCKER_
#else
# define GTEST_AMBIGUOUS_ELSE_BLOCKER_ switch (0) case 0: default:  // NOLINT
#endif
```
and simply doesn't do the job.

---
I tried to correct it by dropping the `GTEST_AMBIGUOUS_ELSE_BLOCKER_` and using instead the
```c++
do { ... } while (false)
```
trick. 

However, this doesn't help either since some of the affected macros allow extra user content after macro invocation like this:
```c++
  if (AlwaysFalse())
    GTEST_CHECK_(false) << "This should never be executed; "
                           "It's a compilation test only.";
```
and the closing `} while (false)` breaks it.

At the moment I don't know how to approach the problem.

**Comment by adambadura** [comment]
I haven't try to implement it yet, just an early idea.

Perhaps we should define those macros as immediately invoked lambdas which return a stream object which either is "the real object" or a "`/dev/null`" object. Something like:
```c++
# define GTEST_CHECK_(condition) \
    []() { \
      if (::testing::internal::IsTrue(condition)) \
        return /dev/null; \
      else \
        return GTEST_LOG_(FATAL) << "Condition " #condition " failed. "; \
    }()
#endif  // !defined(GTEST_CHECK_)
```

**Comment by adambadura** [comment]
Or perhaps somewhat less generic but simpler in this case use of ternary expression:
```c++
# define GTEST_CHECK_(condition) \
    ::testing::internal::IsTrue(condition) \
      ? /* default constructed stream for no-op */
      : GTEST_LOG_(FATAL) << "Condition " #condition " failed. "
```

**Comment by kuzkry** [comment]
You're right, this fails on all GCCs I've got, i.e. versions 7, 8 and 9. You're especially right about `GTEST_AMBIGUOUS_ELSE_BLOCKER_` - this should be removed.
About implementation: Your code sounds like a plan, you could prepare a pull request for that.

Btw. If your solution doesn't work, my first naive thought was that maybe we could implement it using some sort of a helper class, something like `class OutputStreamableOnAssertionFailure` which could be just used in `GTEST_CHECK_`, so that it would internally check the condition and conditionally allow calls to operator<< to provide a failure string. It's an alternative idea you could play around with, but I'd much rather see your version though.

**Comment by adambadura** [comment]
@kuzkry I tried and as for now, I failed.

I have decided to go with the immediately invoked lambda since it is a more generic approach and also seems safer (the ternary operator trick seems to have its own risks). There are however some issues I did not expect and did not solve, yet at least.

Firstly, the lambda has to return something. The first idea is to follow what the `GTEST_LOG_` returns - namely an `::std::ostream&`. However, returning a reference requires a non-temporary object. `GTEST_LOG_` returns `::std::cerr` which is fine, but for the lambda, it is an issue.

I did work around it but only to find out that now linker claims that `Random::kMaxRange` is not defined. Why now? I have no idea... Apparently its use in one of the (now changed) `GTEST_CHECK_` was no longer eradicated to compile-time only. Surprisingly, making it a `constexpr` (which it can and should) didn't help.

Then I have just replaced uses of `kMaxRange` with its definition, the `(1u << 31)`. Only to find out that now some tests started to fail for now apparent reason.

At this point, I'm giving up for now. I will try to improve something in the area and in the meantime get a better understanding of what the code really does. So, if you have some ideas on how to approach this, be my guest! :)

**Comment by gennadiycivil** [comment]
@adambadura Thank you for trying all this. I am not 100% sure how practically important this warning is.

Can you provide a practical reason to try and make the compilation clean with this warning?

Thanks again

**Comment by adambadura** [comment]
@gennadiycivil My preference is to have warning-free compilation while at the same time use a large number of checks.

This is more important for the code used by clients (so that they will never have to work around warnings in the library code). While I do know this case is "private part", but

1. Once you have some code it feels easy and natural to reuse it elsewhere by which a warning may "leak" into "client space". So, there is a risk.
2. A warning prevents you from using `-Werror` or similar, which makes it easier to introduce new warnings.

In this particular case (with the `GTEST_CHECK_` - although to be honest I didn't check all warning cases if they only refer to this macro) I think the warning is a false alarm. The construction seems safe. However, if the code would be warning free we wouldn't have to spend time analyzing the warnings. :)

Also, in this particular case, I think using immediately invoked lambda results in a safer macro. Even if it doesn't matter much for the `GTEST_CHECK_` it does lay ground to other macros where it may matter and it helps develop ourself to seek better solutions.

After sleeping it over I think we should perhaps - at the very least - remove `GTEST_AMBIGUOUS_ELSE_BLOCKER_` since:

1. apparently, it doesn't help,
2. it is pretty wired - I was surprised such code is even valid.

However, other compilers need to be checked first since perhaps it does help there.

**Comment by kuzkry** [comment]
> After sleeping it over I think we should perhaps - at the very least - remove `GTEST_AMBIGUOUS_ELSE_BLOCKER_` since:
> 
>     1. apparently, it doesn't help,
> 
>     2. it is pretty wired - I was surprised such code is even valid.
> 
> 
> However, other compilers need to be checked first since perhaps it does help there.

Doesn't `switch(0) case 0: default:` tell you straight away it's a dummy stuff to mute compiler complains? It's really that trivial. :D Look: https://godbolt.org/z/W5cRy4
You may remove this silly macro.

**Comment by adambadura** [comment]
@kuzkry I did check the same before posting this issue. What I meant is that without seeing it does compile I would say it does not. I was surprised this is syntax is valid.

None the less, it doesn't trick GCC. I'm not sure how about other compilers.

**Comment by EricWF** [comment]
GCC is correct to produce the warning, and the test case is correct to ignore it because the test is trying to verify that even poorly written code works as expected.

The ambiguous else blocker macro does affect both GCC and Clangs diagnostics. Though for GCC it doesn't suppress all cases.
https://godbolt.org/z/mPFmIe

**Comment by EricWF** [comment]
Since the goal of this issue was to reduce the number of warnings emitted, and since more warnings are emitted without the blocker than with it, is there a reason to keep this issue open?

At most I think we should suppress the warning inside `googletest-port-test.cc` when we're specifically testing for ambiguous else problems (which is the diagnostic mentioned at the start).

Outside of those tests, It's my position that the GCC diagnostics are correct, and users should add braces to their code.

**Comment by adambadura** [comment]
@EricWF, if the blocker does reduce the number of warnings emitted (I haven't checked that) then indeed, there is no point in removing it without fixing the warnings in some other way.

However, do I understand it correctly, that the code is purposefully written so that the warning is emitted? Why?
